content/blog/build-publish-containers-iac/index.md

```diff
@@ -0,0 +1,2252 @@
+
+        // Export the base and specific version image name.
+        ctx.Export("baseImageName", image.BaseImageName)
+        ctx.Export("fullImageName", image.ImageName)
+        return nil
+    })
+}
+```
+
+{{< /choosable >}}
+
+{{< choosable language csharp >}}
+
+```csharp
+// [Existing imports ...]
+
+using Pulumi.Docker;
+
+class Program
+{
+    static Task<int> Main() => Deployment.RunAsync(async () => {
+        // [Registry configuration as shown above ...]
+
+        // Build and publish the container image.
+        var image = new Image("my-image", new ImageArgs
+        {
+            Build = new DockerBuild { Context = "app" },
+            ImageName = imageName,
+            Registry = registryInfo,
+        });
+
+        // Export the base and specific version image name.
+        return new Dictionary<string, object>
+        {
+            { "baseImageName", image.BaseImageName },
+            { "fullImageName", image.ImageName },
+        };
+    });
+}
+```
+
+{{< /choosable >}}
+
+Now let's run `pulumi up`. If you haven't already done so, this will provision the cloud resources, as well as build/publish the container image. As the build runs, you will see Docker build output streamed to your terminal:
+
+```bash
+$ pulumi up
+Updating (dev)
+
+     Type                   Name            Status
+ +   pulumi:pulumi:Stack    project-dev     created
+ ... cloud-specific resources omitted ...
+ +   └─ docker:image:Image  my-image        created
+
+Outputs:
+    baseImageName: "...cloud-specific url.../my-repo-dc811b0"
+    fullImageName: "...cloud-specific url.../my-repo-dc811b0:78d0fce7c2450c15a6153b6b11208fcb6b9edea7bb7ef3b7b6194f3fc101a170"
+```
+
+Note also the base and versioned image URLs are exported as stack outputs. These are optional but have three benefits:
+
+1. As we see here, the CLI will print them after each deployment.
+2. The CLI can fetch them on-demand, e.g., 'pulumi stack output fullImageName', making it easy to script actress to these container images.
+3. The [Pulumi `StackReference` component]({{< relref "/docs/intro/concepts/organizing-stacks-projects#inter-stack-dependencies" >}}) can be used to depend on this stack from another, allowing up to build higher levels of infrastructure that consume these images.
+
+As an example of (2) in action, let's run the image locally using the exported image name:
+
+```bash
+$ pulumi stack output fullImageName
+...cloud-specific url.../my-repo-dc811b0:78d0fce7c2450c15a6153b6b11208fcb6b9edea7bb7ef3b7b6194f3fc101a170
+```
+
+> Note that this requires that our client can pull from our provisioned registry, which may require cloud-specific authentication first.
+
+To redeploy changes to the `Dockerfile` or anything that image depends on, including application code, just rerun `pulumi up`. It will detect the differences, rebuild and repush the minimal layer changes necessary to update and retag the repository image, and then export the resulting image name.
```

Comment:
```suggestion
To redeploy changes to the `Dockerfile` or anything that image depends on, including application code, rerun `pulumi up`. It will detect the differences, rebuild and re-push the minimal layer changes necessary to update and retag the repository image, and then export the resulting image name.
```